core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -366,7 +396,12 @@ public void onNext(T item) {
         @Override
         public void onError(Throwable cause) {
             requireNonNull(cause, "cause");
-            close(cause);
+            if (finalizer != null) {
+                downstream.onNext(finalizer.apply(cause));
+                close0(null);
+            } else {
+                close(cause);
```

Comment:
I think we can also call `close0(cause)` right away.

Reply:
✅ updated!